### PR TITLE
Fastest version

### PR DESCRIPTION
**Pull request by sonots**
depends on https://github.com/st-tech/teyu/pulls

Before:

```
Warming up --------------------------------------
              Normal     1.180M i/s -      1.235M times in 1.046608s (847.32ns/i)
                teyu   125.081k i/s -    130.812k times in 1.045816s (7.99μs/i)
         attr_extras   226.015k i/s -    232.401k times in 1.028254s (4.42μs/i)
Calculating -------------------------------------
              Normal     1.353M i/s -      3.541M times in 2.616627s (739.04ns/i)
                teyu   153.047k i/s -    375.243k times in 2.451815s (6.53μs/i)
         attr_extras   217.607k i/s -    678.045k times in 3.115912s (4.60μs/i)

Comparison:
              Normal:   1353104.2 i/s
         attr_extras:    217607.2 i/s - 6.22x  slower
                teyu:    153047.0 i/s - 8.84x  slower
```

After:

```
Warming up --------------------------------------
              Normal     1.422M i/s -      1.442M times in 1.013919s (703.23ns/i)
                teyu     1.376M i/s -      1.407M times in 1.022615s (726.84ns/i)
         attr_extras   241.096k i/s -    253.956k times in 1.053340s (4.15μs/i)
Calculating -------------------------------------
              Normal     1.384M i/s -      4.266M times in 3.083453s (722.79ns/i)
                teyu     1.183M i/s -      4.127M times in 3.487852s (845.04ns/i)
         attr_extras   191.217k i/s -    723.287k times in 3.782541s (5.23μs/i)

Comparison:
              Normal:   1383524.6 i/s
                teyu:   1183376.2 i/s - 1.17x  slower
         attr_extras:    191217.2 i/s - 7.24x  slower
```

### CODE REVIEW

File: lib/teyu.rb [hunk 1]
```diff
@@ -15,77 +15,65 @@ def initialize(klass, arg_names)
...
-          instance_variable_set(:"@#{name}", value)
-        end
-
-        keyword_arg_values = arg_values[sorter.req_args.length..].find { |value| value.is_a?(Hash) } || {}
-        (sorter.keyword_args & keyword_arg_values.keys).each do |name|
-          instance_variable_set(:"@#{name}", keyword_arg_values[name])
-        end
-      end
+      definer = Teyu::ArgsDefiner.new(@arg_names)
+      @klass.class_eval(definer.def_initialize, __FILE__, __LINE__ + 1)
```
Comment (sonots) [inline lib/teyu.rb]: The point is that `teyu_init :a, :b!, :c!, :d!, e: 'e'` just generates

```
@klass.class_eval <<~EOS
      def initialize(a, b:, c:, d:, e: 'e')
        @a = a
        @b = b
        @c = c
        @d = d
        @e = e
      end
EOS
```

File: lib/teyu.rb [hunk 2]
```diff
@@ -3,89 +3,71 @@
...
     end
-  end
 
-  class ArgsSorter
-    REQUIRED_SYMBOL = '!'.freeze
+    private def def_initialize_args
+      args = []
+      args << "#{required_positional_args.map(&:to_s).join(', ')}"
+      args << "#{required_keyword_args.map { |arg| "#{arg}:" }.join(', ')}"
+      args << "#{optional_keyword_args.map { |k, v| "#{k}: #{v.inspect}" }.join(', ')}"
```
Comment (sonots) [inline lib/teyu.rb]: This version does not allow

```
teyu_init :a, b: Object.new
```

but fast.

Comment (sonots) [reply lib/teyu.rb]: IDEA NOTE: fallback to the old implementation if `class_eval` version failed.

**Comment by sonots** [comment]
rebased

**Comment by takanamito** [comment]
Thanks!

**Pull request merged**